### PR TITLE
Refactor interceptors code with chain

### PR DESCRIPTION
**Pull request by wy65701436**
1, add a blob inteceptors for quota usage
2, add a manifest inteceptors for quota usage

Signed-off-by: wang yan <wangyan@vmware.com>

### CODE REVIEW

File: src/core/controllers/controllers_test.go [hunk 1]
```diff
@@ -32,7 +32,7 @@ import (
 	"github.com/goharbor/harbor/src/common/models"
 	utilstest "github.com/goharbor/harbor/src/common/utils/test"
 	"github.com/goharbor/harbor/src/core/config"
-	"github.com/goharbor/harbor/src/core/proxy"
+	"github.com/goharbor/review/harbor/src/core/proxy"
```
Comment (heww) [inline src/core/controllers/controllers_test.go]: This import path may not exist. Should it be `github.com/goharbor/src/core/proxy` ?

File: src/core/middlewares/chain.go [hunk 2]
```diff
@@ -0,0 +1,126 @@
...
+				return nil
+			}
+			return constructor(next)
+		})
+	}
+	return &chain
+}
+
+func (b *DefaultCreator) getMiddleware(mName string) alice.Constructor {
+	var middleware alice.Constructor
```
Comment (heww) [inline src/core/middlewares/chain.go]: This `getMiddleware` method can be implemented using table-driven methods.

```go
middlewares := map[string]alice.Constructor {
  READONLY: func(next http.Handler) http.Handler { return readonly.New(next) },
  URL: func(next http.Handler) http.Handler { return url.New(next) },
  // ...
}

return middlewares[mName]
```

`middlewares` can be defined and initialized outside this method for performance.

File: src/core/controllers/controllers_test.go (hunk 1 above)
Comment (wy65701436) [reply src/core/controllers/controllers_test.go]: ok thanks.

File: src/core/middlewares/regquota/handler.go [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+}
+
+// ServeHTTP PATCH manifest ...
+func (rqh regQuotaHandler) ServeHTTP(rw http.ResponseWriter, req *http.Request) {
+	match, _, _ := util.MatchManifestURL(req)
+	if match {
+		var mfSize int64
+		var mfDigest string
+		mediaType := req.Header.Get("Content-Type")
+		if req.Method == http.MethodPut && mediaType == "application/vnd.docker.distribution.manifest.v2+json" {
```
Comment (ywk253100) [inline src/core/middlewares/regquota/handler.go]: There are some other media types that should be supported, refer to  https://github.com/goharbor/harbor/blob/master/src/replication/transfer/image/transfer.go#L263

File: src/core/middlewares/util/util.go [hunk 4]
```diff
@@ -0,0 +1,168 @@
...
+	if err != nil {
+		log.Debugf("failed to marshal json error, %v", err)
+		return msg
+	}
+	return string(str)
+}
+
+// MatchManifestURL ...
+func MatchManifestURL(req *http.Request) (bool, string, string) {
+	re := regexp.MustCompile(manifestURLPattern)
```
Comment (ywk253100) [inline src/core/middlewares/util/util.go]: Should use `Compile` instead, the `MustCompile` panics if the pattern cannot be parsed.

**Pull request merged**